### PR TITLE
Qi readme

### PR DESCRIPTION
**Pull request by Idan-BenAmi**

### CODE REVIEW

File: model_compression_toolkit/quantizers_infrastructure/README.md [hunk 1]
```diff
@@ -0,0 +1,93 @@
+# Quantizers Infrastructure (QI)
```
Comment (elad-c) [inline model_compression_toolkit/quantizers_infrastructure/README.md:1]: Need license?

File: model_compression_toolkit/quantizers_infrastructure/README.md [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+Quantizers infrastructure is a library that provides quantization modules for hardware-oriented model optimization tools.
+
+The quantization modules (quantizers) can be used for emulating inference-time quantization (inferable-quantizers) and for optimization of model quantization during training (trainable quantizers).
+
+## Library structure
+
+### Quantizer
+
+Quantizer instance implements the quantization function: conversion of floating point values to quantized values represented in floating point (fake-quantization).
```
Comment (elad-c) [inline model_compression_toolkit/quantizers_infrastructure/README.md:11]: 1. A\The
2. The quantizer quantizes an input tensor that might be float (weights) or quantized (most activations)

File: model_compression_toolkit/quantizers_infrastructure/README.md [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+Quantizer instance should be defined per layer, separately for weights quantization and activations quantization.
```
Comment (elad-c) [inline model_compression_toolkit/quantizers_infrastructure/README.md:12]: A\The

File: model_compression_toolkit/quantizers_infrastructure/README.md [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+The quantizers are divided into two main types:
+
+`BaseInferableQuantizer` - A type of quantizer that uses for emulating inference-time quantization
```
Comment (elad-c) [inline model_compression_toolkit/quantizers_infrastructure/README.md:16]: is used for

File: model_compression_toolkit/quantizers_infrastructure/README.md [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+### Quantizer
+
+Quantizer instance implements the quantization function: conversion of floating point values to quantized values represented in floating point (fake-quantization).
+Quantizer instance should be defined per layer, separately for weights quantization and activations quantization.
+
+The quantizers are divided into two main types:
+
+`BaseInferableQuantizer` - A type of quantizer that uses for emulating inference-time quantization   
+
+`BaseTrainableQuantizer` - A type of quantizer that uses for quantization optimization during training. This type of quantizer contains learnable parameters
```
Comment (elad-c) [inline model_compression_toolkit/quantizers_infrastructure/README.md:18]: is used for

File: model_compression_toolkit/quantizers_infrastructure/README.md [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+Usage example:
+
+```python
+from model_compression_toolkit import quantizers_infrastructure as qi
+
+# generate a signed 8-bits symmetric quantizer for numbers in the interval [-2,2)
```
Comment (elad-c) [inline model_compression_toolkit/quantizers_infrastructure/README.md:25]: interval [-2, 2-LSB]

File: model_compression_toolkit/quantizers_infrastructure/README.md [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+# generate a signed 8-bits symmetric quantizer for numbers in the interval [-2,2) 
+w_quantizer = qi.WeightsSymmetricInferableQuantizer(num_bits=8,
+                                                    threshold=2,
+                                                    signed=True,
+                                                    per_channel=True)
+```
+### Dispatcher
+
+`NodeQuantizationDispatcher` instance contains a list of relevant quantizers for node (layer) in the model:
```
Comment (elad-c) [inline model_compression_toolkit/quantizers_infrastructure/README.md:33]: The NodeQuantizationDispatcher`
a node

File: model_compression_toolkit/quantizers_infrastructure/README.md [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+weights quantizer per attribute (kernel, bias, ...)
+activation quantizer per layer's output
```
Comment (elad-c) [inline model_compression_toolkit/quantizers_infrastructure/README.md:35]: (Add bullets)

_The_ weights quantizer _quantizes a tensor_ per attribute_\(maybe per channel?)_ (kernel, bias, ...)
_The_ activation quantizer _quantizes a tensor_ per layer's output

File: model_compression_toolkit/quantizers_infrastructure/README.md [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+
+Usage example:
+
+```python
+# generate node dispatcher with the above weights quantizer
```
Comment (elad-c) [inline model_compression_toolkit/quantizers_infrastructure/README.md:40]: a node

File: model_compression_toolkit/quantizers_infrastructure/README.md [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+
+`NodeQuantizationDispatcher` instance contains a list of relevant quantizers for node (layer) in the model:
+weights quantizer per attribute (kernel, bias, ...)
+activation quantizer per layer's output
+
+Usage example:
+
+```python
+# generate node dispatcher with the above weights quantizer 
+dispatcher = qi.KerasNodeQuantizationDispatcher(w_quantizer)
```
Comment (elad-c) [inline model_compression_toolkit/quantizers_infrastructure/README.md:41]: The dispatcher is called with a dict: {name: quantizer}
also, use the kwarg when calling a class: qi.KerasNodeQuantizationDispatcher(weight_quantizers=your dict)

File: model_compression_toolkit/quantizers_infrastructure/README.md [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+```
+### Quantization Wrapper
+
+A quantization wrapper instance is used as a layer's replacement which emulates the layer's operation only with quantization.
```
Comment (elad-c) [inline model_compression_toolkit/quantizers_infrastructure/README.md:45]: remove "only"

File: model_compression_toolkit/quantizers_infrastructure/README.md [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+The quantization wrapper receives the layer to wrap, and a dispatcher for mapping the relevant quantizer per attribute. 
+
+`KerasQuantizationWrapper` - used to wrap Keras layer
```
Comment (elad-c) [inline model_compression_toolkit/quantizers_infrastructure/README.md:48]: "to wrap" => for warpping a

File: model_compression_toolkit/quantizers_infrastructure/README.md [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+
+`PytorchQuantizationWrapper` - used to wrap torch nn module
```
Comment (elad-c) [inline model_compression_toolkit/quantizers_infrastructure/README.md:50]: a torch

File: model_compression_toolkit/quantizers_infrastructure/README.md [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+qi.KerasQuantizationWrapper(layer=model.layers[1], dispatcher)
+```
+
+### Quantization configuration
+
+Quantizer configuration or quantizer parameters can be stored in the `BaseQuantizerConfig` object. 
+
+For trainable quantizer we use:
+
+`TrainableQuantizerWeightsConfig` - contains the configuration of weight quantization for the trainable quantizer.
```
Comment (elad-c) [inline model_compression_toolkit/quantizers_infrastructure/README.md:72]: the trainable quantizer => a trainable quantizer

File: model_compression_toolkit/quantizers_infrastructure/README.md [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+
+`TrainableQuantizerActivationConfig` - contains the configuration of activation quantization for trainable quantizer
```
Comment (elad-c) [inline model_compression_toolkit/quantizers_infrastructure/README.md:74]: the trainable quantizer => a trainable quantizer

File: model_compression_toolkit/quantizers_infrastructure/README.md [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+
+For inferable quantizers we don't use the `BaseQuantizerConfig` object, instead, we use explicit parameters per quantizer  
+
+Usage example:
+```python
+# generate example configurations for trainable quantizer
```
Comment (elad-c) [inline model_compression_toolkit/quantizers_infrastructure/README.md:80]: for a

File: model_compression_toolkit/quantizers_infrastructure/README.md [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+q_config = qi.TrainableQuantizerWeightsConfig(weights_quantization_method=QuantizationMethod.POWER_OF_TWO,
```
Comment (elad-c) [inline model_compression_toolkit/quantizers_infrastructure/README.md:81]: import for QuantizationMethod

File: model_compression_toolkit/quantizers_infrastructure/README.md [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+```python
+from tensorflow import keras
+
+# generate an example of Keras model
+inputs = keras.layers.Input(shape=(32,32,3))
+x = keras.layers.Conv2D(6, 7, use_bias=False)(inputs)
+model = keras.Model(inputs=inputs, outputs=x)
+
+# create quantization wrapper from the convolution layer according to the dispatcher
+qi.KerasQuantizationWrapper(layer=model.layers[1], dispatcher)
```
Comment (elad-c) [inline model_compression_toolkit/quantizers_infrastructure/README.md:63]: Don't we want the "clone" code for actually replacing the layer with the wrapped layer?

**Review by elad-c — CHANGES_REQUESTED** [review]

File: model_compression_toolkit/quantizers_infrastructure/README.md [hunk 19]
```diff
@@ -0,0 +1,34 @@
+# Quantizers Infrastructure (QI)
+
+Quantizers infrastructure is a module containing quantization abstraction and quantizers for hardware-oriented model optimization tools such as the Model Compression Toolkit ([MCT](https://github.com/sony/model_optimization)).
+
+It provides the required abstraction for emulating inference-time quantization and trainable quantization methods such as quantization-aware training.
+
+## High level description
+
+To create quantization abstraction in a model,
```
Comment (elad-c) [inline model_compression_toolkit/quantizers_infrastructure/README.md:9]: why start a new line after the comment?

File: model_compression_toolkit/quantizers_infrastructure/README.md [hunk 20]
```diff
@@ -0,0 +1,34 @@
...
+we replace each layer with a "Quantization Wrapper" - a more complex layer, which includes the original layer, a weights quantizer block, and activations quantizer block. A quantization dispatcher is attached to each quantization wrapper to set the quantizer type.
+
+<img src="../../docsrc/images/quantization_infra.png" width="700">
+
+The quantizers in this module are divided into two main types:
+"Inferable Quantizer", is used for emulating inference-time quantization, and "Trainable Quantizer", contains learnable quantization parameters that can be optimized during training.
```
Comment (elad-c) [inline model_compression_toolkit/quantizers_infrastructure/README.md:15]: The "Inferable Quantizer" is used

**Review by elad-c — APPROVED** [review]
few minor comments

**Review by reuvenperetz — APPROVED** [review]

**Pull request merged**